### ISSUE TITLE
[start here] learn scientific execution infrastructure: bounded containers, durable jobs, and trustworthy results

### ISSUE THREAD

**Issue by haidmoham** [haidmoham/lmlab#5]
## priority and outcome

**start here — first of two linked learning projects.** build the execution foundation in lmlab, then use spider's checkpoint evaluations as its second workload. do not build two separate schedulers.

researcher-facing goal: submit an existing, bounded language-model comparison; leave it running; inspect progress; recover from a worker failure; and obtain a result whose code, data, experimental protocol, and completion status can be trusted.

this develops containers, isolation, durable execution, measured concurrency, and scientific product engineering through work the lab already needs. it is not a new model implementation, a commercial platform, or a promise of hiring eligibility.

## why this repository first

reviewed `main` at `d7e07f74a97a7a6c8b3809eeefdf913d061758b3` by source inspection; no experiments or tests were executed for this issue.

- [`compare_bigram_transformer.py`](https://github.com/haidmoham/lmlab/blob/d7e07f74a97a7a6c8b3809eeefdf913d061758b3/experiments/compare_bigram_transformer.py) already exposes `run(steps, seeds, output, ...)`, matched initialization, fixed evaluation windows, final/full checkpoints, source hashes, and timing. it is a small CPU workload with no simulator/viewer dependency.
- [`compare_context_lengths.py`](https://github.com/haidmoham/lmlab/blob/d7e07f74a97a7a6c8b3809eeefdf913d061758b3/experiments/compare_context_lengths.py) and [`test_context_comparison.py`](https://github.com/haidmoham/lmlab/blob/d7e07f74a97a7a6c8b3809eeefdf913d061758b3/tests/test_context_comparison.py) establish useful scientific invariants: matched prediction targets, independent parameter objects, and explicit comparison protocols.
- these runners currently use local paths, process-global torch state, and direct artifact writes. they reject reused output directories; they are not yet durable job services.
- [#3](https://github.com/haidmoham/lmlab/issues/3) already proposes distributed experiment execution. this issue supplies a smaller implementation path using an existing experiment; cross-lingual research is not a prerequisite. reuse the resulting infrastructure there rather than implementing its systems checklist twice.
- [#4](https://github.com/haidmoham/lmlab/issues/4) remains the separate wafer GPU curriculum. do not turn this issue into transformer optimization or make CUDA completion a prerequisite.

spider already has richer domain artifacts, but its different controller/checkpoint schemas, candidate physics, and replay semantics add more variables. bring those in after basic execution/failure semantics work.

## learning ownership

follow [AGENTS.md](https://github.com/haidmoham/lmlab/blob/d7e07f74a97a7a6c8b3809eeefdf913d061758b3/AGENTS.md). **the learning target now includes infrastructure**, so the agent must not silently implement the entire runner under the old label of routine plumbing.

human-owned work: job identity, state transitions, process/container lifecycle, isolation choices, completion protocol, retry/lease semantics, failure diagnosis, and interpretation of measurements. agent help: explain mechanisms, prepare stubs and environments, remove incidental friction, review code, and automate agreed repetitive checks. use assignment-sized increments, not line-by-line quizzes. a working agent-generated service alone is not evidence of understanding.

for each slice, retain a short record: expectation → implementation attempt → measurement/failure → explanation → next change. unresolved results are valid records. public progress notes should describe the engineering, not private correspondence.

## slice 1 — one inspectable local job

**question:** what must travel with a command for someone else to know what experiment actually ran?

- [ ] inspect the existing `run()` API and make one short fresh baseline, subject to the normal execution agreement. an existing bounded entry point is:

```bash
uv sync
uv run python -c "from experiments.compare_bigram_transformer import run; run(steps=10, seeds=(42,), output='artifacts/infra-smoke-local')"
```

this is a plumbing smoke test, not a scientific training result. it requires `data/tiny_shakespeare.txt` and the saved `data/tiny_shakespeare_bpe.json`; `uv sync` does not supply those. identify/reproduce their preparation explicitly and preserve their hashes. choose a fresh output directory on subsequent executions.

- [ ] write a small validated JSON job specification and an adapter around the existing function. initially, **one job is the whole paired comparison for one seed**, not separate bigram/transformer tasks. preserve shared initialization and batch pairing.
- [ ] distinguish stable experiment/job identity from unique execution-attempt identity. include source revision/content, dataset/tokenizer hashes, seed, treatment/protocol, budget, and environment/execution profile. timestamps and attempt IDs are not scientific identity.
- [ ] validate allowed workload names and parameters before launching anything; no arbitrary shell commands or module paths from user input.
- [ ] capture stdout/stderr, exit status, elapsed time, and expected result/checkpoint files. execute jobs in separate processes rather than concurrent threads sharing torch RNG state.

**evidence:** a new directory contains the resolved spec, attempt record, logs, and parseable result. missing data produces a useful error rather than a mysteriously failed training job.

### scientific trap to preserve

[`attention-resume.md`](https://github.com/haidmoham/lmlab/blob/d7e07f74a97a7a6c8b3809eeefdf913d061758b3/experiments/attention-resume.md) says historical saved experiments used **pre-norm**, while current reference code is **post-norm**. matching state-dict keys do not imply matching computation. use a fresh same-revision control for this issue, or explicitly recover the historical source when reproducing historical results. do not silently compare the two. record tokenizer identity as well as corpus identity.

## slice 2 — contain that job without changing its meaning

**question:** which properties come from an image, which from kernel isolation, and which from the supervisor?

- [ ] build a pinned, documented Python/CPU environment and run the same spec inside it. verify the host's container runtime and limit support rather than assuming laptop and Linux-host behavior are interchangeable.
- [ ] resolve the current `git rev-parse HEAD` dependency deliberately: capture source provenance at build/submission time and make it available to the runner without mounting a credential-bearing `.git` directory or inventing a commit. preserve source-file hashes.
- [ ] make inputs read-only and outputs attempt-specific. retain only the required writable scratch/output locations. bound CPU, memory, process count, elapsed time, and log/artifact volume; verify limits actually take effect. record torch thread count with the CPU budget.
- [ ] use non-root execution, dropped unnecessary capabilities, no-new-privileges, the default seccomp policy, and no network during the computation once inputs are staged. do not mount host home directories, secrets, or the Docker socket into a workload; do not use privileged containers.
- [ ] exercise bounded benign fixtures for timeout, denied writes, resource exhaustion, and excessive output. preserve the failure reason; cleanup must not erase the evidence.
- [ ] compare native/container scientific outputs with the same declared protocol. choose and report numerical tolerances; do not require file-byte equality or promise bitwise reproducibility across hardware/PyTorch versions.

**scope:** this is an allowlisted, trusted-workload runner with tested containment. containers/resource limits are not proof of a secure service for arbitrary hostile Python. write down the threat model and what is not isolated. rootless runtime feasibility and stronger isolation are explicit design questions, not implied guarantees.

**evidence:** one good job and several safely bounded failure fixtures, with their effective limits and observed outcomes.

## slice 3 — durable jobs, then bounded parallelism

**question:** what remains true when the process that remembers the work disappears?

- [ ] start with a single-host coordinator and local durable state, e.g. SQLite, plus independent container attempts. keep write transactions short; do not hold a database transaction while training. do not share a SQLite file across remote hosts.
- [ ] define a small explicit state machine: queued, running, succeeded, failed, timed_out, cancelled, with retry transitions where justified. persist attempts, leases/heartbeats, exit reasons, and retry counts.
- [ ] atomically claim work. use attempt/lease ownership when renewing or completing it so a stale worker cannot overwrite a newer accepted result.
- [ ] add a bounded worker pool and queue/backpressure. jobs submit only through the coordinator; workload containers do not receive database/daemon credentials.
- [ ] stage artifacts in immutable attempt directories and publish an accepted result only after validating expected files, configuration, and completeness. define how the database pointer and filesystem artifacts recover when only one side is committed.

**important current-code detail:** `results.json` is updated after each seed, and checkpoint files are written directly. a directory or parseable JSON file is not itself a success marker. the adapter must know the expected run set. do not have concurrent workers append to a shared `results.json`.

**evidence:** restart the coordinator and recover queued/running work. submitting the same spec twice does not create two scientific results.

## slice 4 — deliberately break completion

**question:** can execution occur more than once while the scientific result is accepted only once?

- [ ] kill a worker during computation; reclaim expired work with a bounded retry policy.
- [ ] kill after artifacts are staged but before acknowledgement. retry without publishing duplicate results or corrupting the successful attempt.
- [ ] delay an old attempt beyond lease expiry and let it finish after a replacement. prove its stale completion cannot win.
- [ ] test cancellation, invalid inputs, partial checkpoint/result writes, and coordinator restart. distinguish retryable execution failure from deterministic input/scientific validation failure.
- [ ] keep crash fixtures small. initially restart short jobs from the beginning; do not pretend saved optimizer/RNG state means a resume feature is already implemented. exact training continuation is a separate, explicitly tested extension.

**invariant:** retries may cause repeated computation, but each job has at most one accepted, complete result. failed/abandoned attempts remain inspectable. do not claim exactly-once execution.

**handoff:** once bounded container execution, durable state, and these recovery tests work, spider can become the second workload. no cloud deployment or Kubernetes prerequisite.

## slice 5 — make the system useful to another researcher

**question:** can someone operate it without understanding the coordinator implementation?

- [ ] expose submit/status/logs/results/cancel through a small CLI plus an inspectable notebook or minimal local interface. show provenance, failure reasons, and partial versus accepted results; keep arbitrary-code upload out of scope.
- [ ] recover the existing scientific presentation: comparable metric table, matched-axis loss curves, and fixed-prompt samples. refuse or clearly separate incompatible protocols, source versions, or tokenizer identities. do not aggregate duplicate retries as extra seeds.
- [ ] have another person, or a genuinely fresh environment, submit a small provided job and reproduce its comparison without editing hidden paths. record friction discovered.
- [ ] measure the same bounded workload set with 1, 2, and, when resources permit, 4 workers. report machine/runtime/thread budgets, startup time, queue wait, wall time, throughput, memory, failures, and correctness. slower execution is an informative finding.
- [ ] separate fast synthetic queue/failure tests from real model-job measurements. do not run 100 expensive training jobs to obtain a scalability screenshot or describe a local worker-pool test as internet scale.

## minimal shape and stop line

start with a thin workload adapter, job spec, supervisor, durable state, and focused tests in the existing repository layout. add abstractions only after repetition demonstrates the need. keep generated runs/weights/logs untracked; check in a small evidence note and reproducible fixture/config.

no new model, distributed model training, Kubernetes, Ray/Celery, billing, public multi-tenancy, custom container runtime, or GPU scheduling in the initial deliverable. the [gVisor security model](https://gvisor.dev/docs/architecture_guide/security/) is a useful follow-up for deciding what stronger isolation would address; implementing it is not a prerequisite for completing this trusted-workload version.

## completion evidence

- [ ] existing tests still pass; record exact checks and any environmental limitations.
- [ ] one native/container comparison with explicit provenance and numerical checks.
- [ ] a failure-injection transcript showing restart, duplicate/stale completion, cancellation, and incomplete-output handling.
- [ ] one measured bounded-concurrency comparison, including negative findings.
- [ ] one user-facing submission/result path and a fresh-user/environment trial.
- [ ] short architecture/threat-model note explaining the guarantees and non-guarantees.
- [ ] a companion spider issue reuses this execution contract rather than duplicating it.

## primary references, read at the relevant slice

- [Docker resource constraints](https://docs.docker.com/engine/containers/resource_constraints/) and [container run options](https://docs.docker.com/engine/containers/run/): effective resource limits and execution configuration.
- [Docker security](https://docs.docker.com/engine/security/), [rootless mode](https://docs.docker.com/engine/security/rootless/), and [seccomp](https://docs.docker.com/engine/security/seccomp/): privilege/isolation boundaries; do not disable protections to hide a setup problem.
- [SQLite transactions](https://www.sqlite.org/lang_transaction.html): atomic state changes and single-writer behavior.
- [PyTorch reproducibility](https://docs.pytorch.org/docs/stable/notes/randomness.html): seeds, environment, and limits of equivalence claims.
